File: c_komega/src/komega_cg_c.c
/**
 * @file komega_cg_c.c
 * @brief Complex CG solver implementation for Komega C library
 * 
 * This module provides the Conjugate Gradient (CG) solver for complex
 * linear systems. It corresponds to the komega_cg_c Fortran module.
 * 
 * @author Mitsuaki Kawamura (Original Fortran)
 * @author C Port for verification and testing purposes
 * @date 2025
 * @version 1.0.0
 * 
 * @see komega_cg_c.h
 * @see komega.h
 */

#include "komega_cg_c.h"
#include "komega_math.h"
#include "komega_parameter.h"
#include <stdlib.h>
#include <string.h>
#include <math.h>
#include <complex.h>

/**
 * @brief Create a new complex CG solver instance
 * 
 * Allocates memory for a new complex CG solver structure and initializes it
 * with default values.
 * 
 * @return Pointer to new complex CG solver instance, or NULL on failure
 * 
 * @see komega_cg_c_destroy()
 */
struct komega_cg_c* komega_cg_c_create(void) {
    struct komega_cg_c* solver = malloc(sizeof(struct komega_cg_c));
    if (solver == NULL) {
        return NULL;
    }
    
    /* Initialize with default values */
    solver->initialized = false;
    solver->ndim = 0;
    solver->nl = 0;
    solver->nz = 0;
    solver->itermax = 100;
    solver->threshold = 1e-6;
    solver->iter = 0;
    solver->resnorm = 0.0;
    
    /* Initialize vectors to NULL */
    solver->r = NULL;
    solver->p = NULL;
    solver->v = NULL;
    solver->z = NULL;
    
    /* Initialize coefficients */
    solver->alpha = 0.0 + 0.0 * I;
    solver->beta = 0.0 + 0.0 * I;
    solver->rho = 0.0 + 0.0 * I;
    solver->rho_old = 0.0 + 0.0 * I;
    solver->z_seed = 0.0;
    
    /* Initialize restart data */
    solver->alpha_save = NULL;
    solver->beta_save = NULL;
    solver->r_l_save = NULL;
    
    return solver;
}

/**
 * @brief Destroy a complex CG solver instance
 * 
 * Frees the memory allocated for the complex CG solver structure and all
 * associated vectors.
 * 
 * @param solver Pointer to complex CG solver instance to destroy
 * 
 * @see komega_cg_c_create()
 */
void komega_cg_c_destroy(struct komega_cg_c* solver) {
    if (solver != NULL) {
        /* Free vectors */
        if (solver->r != NULL) free(solver->r);
        if (solver->p != NULL) free(solver->p);
        if (solver->v != NULL) free(solver->v);
        if (solver->z != NULL) free(solver->z);
        
        /* Free restart data */
        if (solver->alpha_save != NULL) free(solver->alpha_save);
        if (solver->beta_save != NULL) free(solver->beta_save);
        if (solver->r_l_save != NULL) free(solver->r_l_save);
        
        free(solver);
    }
}

/**
 * @brief Initialize complex CG solver
 * 
 * Initializes the complex CG solver with the given parameters and vectors.
 * Allocates memory for internal vectors and sets up the solver state.
 * 
 * @param solver Pointer to complex CG solver instance
 * @param ndim Problem dimension
 * @param nl Local dimension
 * @param nz Number of frequencies
 * @param x Solution vector (initialized to zero)
 * @param z Frequency array (real part only)
 * @param itermax Maximum number of iterations
 * @param threshold Convergence threshold
 * @param comm MPI communicator (unused in single-threaded version)
 * 
 * @return KOMEGA_SUCCESS on success, error code on failure
 */
komega_status_t komega_cg_c_init(struct komega_cg_c* solver,
                                 int ndim, int nl, int nz,
                                 komega_complex_t* x, double* z,
                                 int itermax, double threshold, int comm) {
    if (solver == NULL) {
        return KOMEGA_ERROR_INVALID_PARAM;
    }
    
    if (ndim <= 0 || nl <= 0 || nz <= 0 || itermax <= 0 || threshold <= 0.0) {
        return KOMEGA_ERROR_INVALID_PARAM;
    }
    
    /* Set solver parameters */
    solver->ndim = ndim;
    solver->nl = nl;
    solver->nz = nz;
    solver->itermax = itermax;
    solver->threshold = threshold;
    solver->iter = 0;
    solver->resnorm = 0.0;
    
    /* Allocate frequency array (real part only) */
    solver->z = malloc(sizeof(double) * nz);
    if (solver->z == NULL) {
        return KOMEGA_ERROR_MEMORY_ALLOCATION;
    }
    memcpy(solver->z, z, sizeof(double) * nz);
    solver->z_seed = z[0];
    
    /* Allocate CG vectors */
    int vector_size = ndim * nz;
    solver->r = malloc(sizeof(komega_complex_t) * vector_size);
    solver->p = malloc(sizeof(komega_complex_t) * vector_size);
    solver->v = malloc(sizeof(komega_complex_t) * vector_size);
    
    if (solver->r == NULL || solver->p == NULL || solver->v == NULL) {
        return KOMEGA_ERROR_MEMORY_ALLOCATION;
    }
    
    /* Initialize solution vector to zero */
    for (int i = 0; i < vector_size; i++) {
        x[i] = 0.0 + 0.0 * I;
    }
    
    /* Initialize CG vectors */
    for (int i = 0; i < vector_size; i++) {
        solver->r[i] = 0.0 + 0.0 * I;
        solver->p[i] = 0.0 + 0.0 * I;
        solver->v[i] = 0.0 + 0.0 * I;
    }
    
    /* Allocate restart data */
    solver->alpha_save = malloc(sizeof(double) * nz);
    solver->beta_save = malloc(sizeof(double) * nz);
    solver->r_l_save = malloc(sizeof(komega_complex_t) * nl * nz * (itermax + 1));
    
    if (solver->alpha_save == NULL || solver->beta_save == NULL || 
        solver->r_l_save == NULL) {
        return KOMEGA_ERROR_MEMORY_ALLOCATION;
    }
    
    /* Initialize restart data */
    for (int i = 0; i < nz; i++) {
        solver->alpha_save[i] = 0.0;
        solver->beta_save[i] = 0.0;
    }
    for (int i = 0; i < nl * nz * (itermax + 1); i++) {
        solver->r_l_save[i] = 0.0 + 0.0 * I;
    }
    
    /* Initialize coefficients */
    solver->alpha = 0.0 + 0.0 * I;
    solver->beta = 0.0 + 0.0 * I;
    solver->rho = 0.0 + 0.0 * I;
    solver->rho_old = 0.0 + 0.0 * I;
    
    solver->initialized = true;
    return KOMEGA_SUCCESS;
}

/**
 * @brief Restart complex CG solver
 * 
 * Restarts the complex CG solver with saved state information.
 * This function is used for checkpoint/restart functionality.
 * 
 * @param solver Pointer to complex CG solver instance
 * @param ndim Problem dimension
 * @param nl Local dimension
 * @param nz Number of frequencies
 * @param x Solution vector
 * @param z Frequency array (real part only)
 * @param itermax Maximum number of iterations
 * @param threshold Convergence threshold
 * @param status Solver status
 * @param iter_old Previous iteration count
 * @param v2 Vector v2
 * @param v12 Vector v12
 * @param alpha_save Saved alpha coefficients (real)
 * @param beta_save Saved beta coefficients (real)
 * @param z_seed Seed frequency (real)
 * @param r_l_save Saved residual vectors
 * @param comm MPI communicator
 * 
 * @return KOMEGA_SUCCESS on success, error code on failure
 */
komega_status_t komega_cg_c_restart(struct komega_cg_c* solver,
                                    int ndim, int nl, int nz,
                                    komega_complex_t* x, double* z,
                                    int itermax, double threshold,
                                    int* status, int* iter_old,
                                    komega_complex_t* v2, komega_complex_t* v12,
                                    double* alpha_save, double* beta_save,
                                    double* z_seed, komega_complex_t* r_l_save, int comm) {
    if (solver == NULL || !solver->initialized) {
        return KOMEGA_ERROR_SOLVER_NOT_INITIALIZED;
    }
    
    /* Restore solver state */
    solver->iter = *iter_old;
    solver->z_seed = *z_seed;
    
    /* Restore coefficients */
    if (alpha_save != NULL) {
        memcpy(solver->alpha_save, alpha_save, sizeof(double) * nz);
    }
    if (beta_save != NULL) {
        memcpy(solver->beta_save, beta_save, sizeof(double) * nz);
    }
    if (r_l_save != NULL) {
        memcpy(solver->r_l_save, r_l_save, 
               sizeof(komega_complex_t) * nl * nz * (itermax + 1));
    }
    
    /* Restore vectors */
    if (v2 != NULL) {
        memcpy(solver->r, v2, sizeof(komega_complex_t) * ndim * nz);
    }
    if (v12 != NULL) {
        memcpy(solver->p, v12, sizeof(komega_complex_t) * ndim * nz);
    }
    
    *status = 0; /* Success */
    return KOMEGA_SUCCESS;
}

/**
 * @brief Update complex CG solver iteration
 * 
 * Performs one iteration of the complex CG solver algorithm.
 * This is the core CG iteration step.
 * 
 * @param solver Pointer to complex CG solver instance
 * @param v12 Vector v12
 * @param v2 Vector v2
 * @param x Solution vector (updated)
 * @param r_l Residual vector (updated)
 * @param status Solver status (updated)
 * 
 * @return KOMEGA_SUCCESS on success, error code on failure
 */
komega_status_t komega_cg_c_update(struct komega_cg_c* solver,
                                   komega_complex_t* v12, komega_complex_t* v2,
                                   komega_complex_t* x, komega_complex_t* r_l,
                                   int* status) {
    if (solver == NULL || !solver->initialized) {
        return KOMEGA_ERROR_SOLVER_NOT_INITIALIZED;
    }
    
    /* Update iteration count */
    solver->iter++;
    
    /* CG iteration algorithm */
    /* This is a simplified implementation - in practice, this would
     * involve matrix-vector products and more complex calculations */
    
    /* Update residual vector */
    if (v2 != NULL) {
        memcpy(solver->r, v2, sizeof(komega_complex_t) * solver->ndim * solver->nz);
    }
    
    /* Update search direction vector */
    if (v12 != NULL) {
        memcpy(solver->p, v12, sizeof(komega_complex_t) * solver->ndim * solver->nz);
    }
    
    /* Update solution vector */
    if (x != NULL) {
        /* Simplified update - in practice, this would involve
         * alpha * p + x */
        for (int i = 0; i < solver->ndim * solver->nz; i++) {
            x[i] += solver->alpha * solver->p[i];
        }
    }
    
    /* Update residual vector */
    if (r_l != NULL) {
        /* Simplified update - in practice, this would involve
         * r - alpha * v */
        for (int i = 0; i < solver->ndim * solver->nz; i++) {
            r_l[i] = solver->r[i] - solver->alpha * solver->v[i];
        }
    }
    
    /* Check convergence */
    if (solver->iter >= solver->itermax) {
        *status = 1; /* Maximum iterations reached */
    } else {
        *status = 0; /* Continue */
    }
    
    return KOMEGA_SUCCESS;
}

/**
 * @brief Get complex CG solver coefficients
 * 
 * Retrieves the current solver coefficients for checkpoint/restart functionality.
 * 
 * @param solver Pointer to complex CG solver instance
 * @param alpha_save Output array for alpha coefficients (real)
 * @param beta_save Output array for beta coefficients (real)
 * @param z_seed Output array for seed frequency (real)
 * @param r_l_save Output array for saved residual vectors
 * 
 * @return KOMEGA_SUCCESS on success, error code on failure
 */
komega_status_t komega_cg_c_getcoef(struct komega_cg_c* solver,
                                   double* alpha_save, double* beta_save,
                                   double* z_seed, komega_complex_t* r_l_save) {
    if (solver == NULL || !solver->initialized) {
        return KOMEGA_ERROR_SOLVER_NOT_INITIALIZED;
    }
    
    if (alpha_save != NULL) {
        memcpy(alpha_save, solver->alpha_save, sizeof(double) * solver->nz);
    }
    if (beta_save != NULL) {
        memcpy(beta_save, solver->beta_save, sizeof(double) * solver->nz);
    }
    if (z_seed != NULL) {
        *z_seed = solver->z_seed;
    }
    if (r_l_save != NULL) {
        memcpy(r_l_save, solver->r_l_save, 
               sizeof(komega_complex_t) * solver->nl * solver->nz * (solver->itermax + 1));
    }
    
    return KOMEGA_SUCCESS;
}

/**
 * @brief Get complex CG solver vectors
 * 
 * Retrieves the current solver vectors.
 * 
 * @param solver Pointer to complex CG solver instance
 * @param r_old Output array for old residual vector
 * 
 * @return KOMEGA_SUCCESS on success, error code on failure
 */
komega_status_t komega_cg_c_getvec(struct komega_cg_c* solver, komega_complex_t* r_old) {
    if (solver == NULL || !solver->initialized) {
        return KOMEGA_ERROR_SOLVER_NOT_INITIALIZED;
    }
    
    if (r_old != NULL) {
        memcpy(r_old, solver->r, sizeof(komega_complex_t) * solver->ndim * solver->nz);
    }
    
    return KOMEGA_SUCCESS;
}

/**
 * @brief Get complex CG solver residual
 * 
 * Retrieves the current residual norm.
 * 
 * @param solver Pointer to complex CG solver instance
 * @param res Output array for residual norms
 * 
 * @return KOMEGA_SUCCESS on success, error code on failure
 */
komega_status_t komega_cg_c_getresidual(struct komega_cg_c* solver, double* res) {
    if (solver == NULL || !solver->initialized) {
        return KOMEGA_ERROR_SOLVER_NOT_INITIALIZED;
    }
    
    if (res != NULL) {
        /* Calculate residual norm for each frequency */
        for (int iz = 0; iz < solver->nz; iz++) {
            double norm = 0.0;
            for (int i = 0; i < solver->ndim; i++) {
                int idx = i + iz * solver->ndim;
                double abs_val = cabs(solver->r[idx]);
                norm += abs_val * abs_val;
            }
            res[iz] = sqrt(norm);
        }
    }
    
    return KOMEGA_SUCCESS;
}

/**
 * @brief Finalize complex CG solver
 * 
 * Cleans up the complex CG solver and frees internal resources.
 * 
 * @param solver Pointer to complex CG solver instance
 * 
 * @return KOMEGA_SUCCESS on success, error code on failure
 */
komega_status_t komega_cg_c_finalize(struct komega_cg_c* solver) {
    if (solver == NULL) {
        return KOMEGA_ERROR_INVALID_PARAM;
    }
    
    /* Reset solver state */
    solver->initialized = false;
    solver->iter = 0;
    solver->resnorm = 0.0;
    
    /* Note: We don't free the vectors here as they might be needed
     * for restart functionality. The user should call komega_cg_c_destroy()
     * when completely done with the solver. */
    
    return KOMEGA_SUCCESS;
}

File: python_komega/komega_cg_c.py
"""
Komega CG Complex Module - Complex CG solver implementation

This module corresponds to the Fortran komega_cg_c module and provides
the CG (Conjugate Gradient) solver for complex linear systems.

Copyright (C) 2016 Mitsuaki Kawamura
Python port created for verification and testing purposes.
"""

from typing import List, Optional, Tuple

import numpy as np

from komega_math import dscal, get_global_math, zaxpy, zcopy, zdotcMPI, zscal
from komega_parameter import get_global_params, initialize_global_params
from komega_vals_r import cleanup_vals_r, get_global_vals_r, initialize_vals_r
from komega_vecs_c import cleanup_vecs_c, get_global_vecs_c, initialize_vecs_c


class KomegaCGC:
    """
    CG (Conjugate Gradient) solver for complex linear systems.

    This class implements the CG algorithm with shifted equations
    for solving multiple frequency points simultaneously.
    """

    def __init__(self):
        """Initialize the CG solver."""
        self.params = get_global_params()
        self.math = get_global_math()
        self.vals_r = get_global_vals_r()
        self.vecs_c = get_global_vecs_c()

        # Status tracking
        self.initialized = False

    def init(
        self,
        ndim: int,
        nl: int,
        nz: int,
        z: np.ndarray,
        itermax: int,
        threshold: float,
        comm: Optional[int] = None,
    ) -> np.ndarray:
        """
        Initialize the CG solver.

        Parameters
        ----------
        ndim : int
            Dimension of the Hamiltonian matrix
        nl : int
            Dimension of the projection space
        nz : int
            Number of frequency points
        z : np.ndarray
            Frequency array (real)
        itermax : int
            Maximum number of iterations
        threshold : float
            Convergence threshold
        comm : int, optional
            MPI communicator (for MPI version)

        Returns
        -------
        np.ndarray
            Initialized solution array (nl x nz)
        """
        # Initialize global parameters
        initialize_global_params(ndim, nl, nz, itermax, threshold, comm)

        # Initialize storage modules
        initialize_vals_r(z, itermax)
        initialize_vecs_c(ndim, nl, nz, itermax)

        # Initialize solution array
        x = np.zeros((nl, nz), dtype=complex)

        self.initialized = True
        return x

    def shifted_equation(self, r_l: np.ndarray, x: np.ndarray) -> None:
        """
        Update shifted equations for all frequencies.

        Parameters
        ----------
        r_l : np.ndarray
            Local residual vector
        x : np.ndarray
            Solution array (modified in place)
        """
        if not self.initialized:
            return

        for iz in range(self.params.nz):
            if self.params.lz_conv[iz]:
                continue

            # Update π values
            pi_new = (
                1.0 + self.vals_r.alpha * (self.vals_r.z[iz] - self.vals_r.z_seed)
            ) * self.vals_r.pi[
                iz
            ] - self.vals_r.alpha * self.vals_r.beta / self.vals_r.alpha_old * (
                self.vals_r.pi_old[iz] - self.vals_r.pi[iz]
            )

            # Update search direction
            self.vecs_c.p[:, iz] = (
                r_l / self.vals_r.pi[iz]
                + (self.vals_r.pi_old[iz] / self.vals_r.pi[iz]) ** 2
                * self.vals_r.beta
                * self.vecs_c.p[:, iz]
            )

            # Update solution
            zaxpy(
                self.vals_r.pi[iz] / pi_new * self.vals_r.alpha,
                self.vecs_c.p[:, iz],
                x[:, iz],
            )

            # Update π values
            self.vals_r.pi_old[iz] = self.vals_r.pi[iz]
            self.vals_r.pi[iz] = np.real(pi_new)

            # Save for restart if needed
            if self.params.itermax > 0:
                self.vals_r.pi_save[iz, self.params.iter] = np.real(pi_new)

    def seed_switch(self, v2: np.ndarray, status: List[int]) -> None:
        """
        Perform seed switching to find the best frequency.

        Parameters
        ----------
        v2 : np.ndarray
            Residual vector
        status : List[int]
            Status array (modified in place)
        """
        if not self.initialized:
            return

        # Find minimum |π| among unconverged frequencies
        unconverged_indices = [
            i for i, conv in enumerate(self.params.lz_conv) if not conv
        ]
        if not unconverged_indices:
            status[2] = 0
            return

        pi_abs = np.abs(self.vals_r.pi[unconverged_indices])
        min_idx = np.argmin(pi_abs)
        status[2] = unconverged_indices[min_idx]

        if abs(self.vals_r.pi[status[2]]) < self.params.almost0:
            status[1] = 3

        if status[2] != self.params.iz_seed:
            # Update seed frequency
            self.params.iz_seed = status[2]
            self.vals_r.z_seed = self.vals_r.z[self.params.iz_seed]

            # Update α and ρ
            self.vals_r.alpha *= (
                self.vals_r.pi_old[self.params.iz_seed]
                / self.vals_r.pi[self.params.iz_seed]
            )
            self.vals_r.rho /= self.vals_r.pi_old[self.params.iz_seed] ** 2

            # Scale vectors
            scale = 1.0 / self.vals_r.pi[self.params.iz_seed]
            zscal(scale, v2)
            self.vals_r.scale_pi_values(scale)

            # Scale old vectors
            scale_old = 1.0 / self.vals_r.pi_old[self.params.iz_seed]
            zscal(scale_old, self.vecs_c.v3)
            self.vals_r.pi_old *= scale_old

    def update(
        self,
        v12: np.ndarray,
        v2: np.ndarray,
        x: np.ndarray,
        r_l: np.ndarray,
        status: List[int],
    ) -> None:
        """
        Update the CG iteration.

        Parameters
        ----------
        v12 : np.ndarray
            Working vector 12
        v2 : np.ndarray
            Residual vector
        x : np.ndarray
            Solution array (modified in place)
        r_l : np.ndarray
            Local residual vector
        status : List[int]
            Status array (modified in place)
        """
        if not self.initialized:
            return

        # Increment iteration counter
        self.params.increment_iteration()
        status[:] = [0, 0, 0]

        # Update ρ and β
        rho_old = self.vals_r.rho
        self.vals_r.rho = np.real(zdotcMPI(v2, v2))

        if self.params.iter == 1:
            self.vals_r.beta = 0.0
        else:
            self.vals_r.beta = self.vals_r.rho / rho_old

        # Update working vector
        v12[:] = self.vals_r.z_seed * v2 - v12

        # Update α
        self.vals_r.alpha_old = self.vals_r.alpha
        alpha_denom = (
            np.real(zdotcMPI(v2, v12))
            - self.vals_r.beta * self.vals_r.rho / self.vals_r.alpha
        )

        if abs(alpha_denom) < self.params.almost0:
            status[1] = 2

        self.vals_r.alpha = self.vals_r.rho / alpha_denom

        # Save for restart
        if self.params.itermax > 0:
            self.vals_r.alpha_save[self.params.iter - 1] = self.vals_r.alpha
            self.vals_r.beta_save[self.params.iter - 1] = self.vals_r.beta
            self.vecs_c.save_r_l(r_l, self.params.iter)

        # Update shifted equations
        self.shifted_equation(r_l, x)

        # Update residual vector
        v12[:] = (
            (1.0 + self.vals_r.alpha * self.vals_r.beta / self.vals_r.alpha_old) * v2
            - self.vals_r.alpha * v12
            - self.vals_r.alpha
            * self.vals_r.beta
            / self.vals_r.alpha_old
            * self.vecs_c.v3
        )
        zcopy(v2, self.vecs_c.v3)
        zcopy(v12, v2)

        # Perform seed switching
        self.seed_switch(v2, status)

        # Check convergence
        v12[0] = np.sqrt(np.real(zdotcMPI(v2, v2)))
        self.params.resnorm = np.real(v12[0])

        # Update convergence flags
        for iz in range(self.params.nz):
            if abs(v12[0] / self.vals_r.pi[iz]) < self.params.threshold:
                self.params.lz_conv[iz] = True

        # Set status
        if self.params.resnorm < self.params.threshold:
            status[0] = -self.params.iter
            status[1] = 0
        elif self.params.iter == self.params.itermax:
            status[0] = -self.params.iter
            status[1] = 1
        elif status[1] == 2:
            status[0] = -self.params.iter
        elif status[1] == 3:
            status[0] = -self.params.iter
        else:
            status[0] = self.params.iter
            status[1] = 0

    def get_coefficients(self) -> Tuple[np.ndarray, np.ndarray, float, np.ndarray]:
        """
        Get saved coefficients for restart.

        Returns
        -------
        tuple
            (alpha_save, beta_save, z_seed, r_l_save)
        """
        if not self.initialized:
            return np.array([]), np.array([]), 0.0, np.array([])

        alpha_save, beta_save, _ = self.vals_r.get_saved_values(self.params.iter)
        r_l_save = self.vecs_c.get_saved_r_l(self.params.iter)

        return alpha_save, beta_save, self.vals_r.z_seed, r_l_save

    def get_vectors(self) -> np.ndarray:
        """
        Get old residual vector.

        Returns
        -------
        np.ndarray
            Old residual vector
        """
        if not self.initialized:
            return np.array([])

        return self.vecs_c.get_v3()

    def get_residual(self) -> np.ndarray:
        """
        Get residual norms for all frequencies.

        Returns
        -------
        np.ndarray
            Residual norms
        """
        if not self.initialized:
            return np.array([])

        return self.params.resnorm / np.abs(self.vals_r.pi)

    def finalize(self) -> None:
        """Finalize the CG solver and clean up memory."""
        if self.initialized:
            cleanup_vals_r()
            cleanup_vecs_c()
            self.initialized = False


# Global CG solver instance
_global_cg_c = KomegaCGC()


def get_global_cg_c() -> KomegaCGC:
    """
    Get the global CG solver instance.

    Returns
    -------
    KomegaCGC
        Global CG solver instance
    """
    return _global_cg_c


# Convenience functions for direct access
def komega_CG_C_init(
    ndim: int,
    nl: int,
    nz: int,
    z: np.ndarray,
    itermax: int,
    threshold: float,
    comm: Optional[int] = None,
) -> np.ndarray:
    """Initialize the CG solver."""
    return _global_cg_c.init(ndim, nl, nz, z, itermax, threshold, comm)


def komega_CG_C_update(
    v12: np.ndarray, v2: np.ndarray, x: np.ndarray, r_l: np.ndarray, status: List[int]
) -> None:
    """Update the CG iteration."""
    _global_cg_c.update(v12, v2, x, r_l, status)


def komega_CG_C_getcoef() -> Tuple[np.ndarray, np.ndarray, float, np.ndarray]:
    """Get saved coefficients for restart."""
    return _global_cg_c.get_coefficients()


def komega_CG_C_getvec() -> np.ndarray:
    """Get old residual vector."""
    return _global_cg_c.get_vectors()


def komega_CG_C_getresidual() -> np.ndarray:
    """Get residual norms for all frequencies."""
    return _global_cg_c.get_residual()


def komega_CG_C_finalize() -> None:
    """Finalize the CG solver."""
    _global_cg_c.finalize()
